### ISSUE TITLE
Python 3.4 Travis fail due to unmet dependencies

### ISSUE THREAD

**Issue by orbeckst** [MDAnalysis/mdanalysis#2066]
The Python 3.4 tests are all failing now:

```
UnsatisfiableError: The following specifications were found to be in conflict:
  - numpy=1.15 -> *[track_features=blas_openblas]
  - numpy=1.15 -> python[version='>=3.7,<3.8.0a0'] -> tk[version='>=8.6.7,<8.7.0a0']
  - python=3.4
Use "conda info <package>" to see the dependencies for each package.
```

Examples:
- https://travis-ci.com/MDAnalysis/mdanalysis/jobs/144464939
- https://travis-ci.org/MDAnalysis/mdanalysis/jobs/425782569

Do we need to pin numpy or should we deactivate 3.4 tests?

cc: @richardjgowers @jbarnoud @tylerjereddy

**Comment by zemanj** [comment]
We could pin the numpy version here as yet another band-aid. Though, what we really want is something like "take the highest possible version that is compatible with this python version". Is there a sane way to achieve this without doing nasty things like parsing the output of `conda info`?

**Comment by richardjgowers** [comment]
I thought figuring this stuff out was something conda was meant to figure out for itself.  Maybe we're not giving conda enough information initially to make these decisions.  Pinning on the 3.4 build is an easy fix which wil allow PRs to continue though

**Comment by orbeckst** [comment]
We will have to do *something* soon because this holds up merging of all PRs.

> I thought figuring this stuff out was something conda was meant to figure out for itself. Maybe we're not giving conda enough information initially to make these decisions.

Yes, I though so, too. Perhaps this is actually a bug in one of the build recipes (maybe @kain88-de sees immediately what's wrong) or in the ci-helper script (???) but that doesn't help us right now.

What is the last version of numpy that works for us in 3.4? We then pin that version, [using ci-helper's NUMPY_VERSION](https://github.com/astropy/ci-helpers/blob/master/README.md)
```bash
PYTHON_VERSION=3.4 NUMPY_VERSION=1.14 CODECOV="true" SETUP_CMD="${PYTEST_FLAGS} --cov=MDAnalysis" 
```
(or whatever is needed)

**Comment by kain88-de** [comment]
This looks like a conda error. Numpy 1.15 has not been build for python3.4. This is coming up now because the 1.15 version has been published 6 days ago.

Nevertheless, this should be reported to ci-helpers. It uses a default [pin of 1.15](https://github.com/astropy/ci-helpers/blob/b8b8a2d9b317b11ab68eb0846f89da21d33dfd4b/travis/setup_dependencies_common.sh#L37) for numpy as the latest stable version. This pin causes a problem for python3.4 so using a different pinned version is a good solution.

**Comment by orbeckst** [comment]
Should we open an issue with consensus-helpers?

**Comment by orbeckst** [comment]
Should we open an issue with ci-helpers?

**Comment by zemanj** [comment]
For the record:
- The quick fix from PR #2067 is in place
- More thorough investigation needed for what's going wrong with conda / ci-helpers
- Issues regarding error handling in ci-helpers should go into an issue/PR there.

**Comment by orbeckst** [comment]
I removed the critical label because we have at least the band-aid #2067 .

**Comment by zemanj** [comment]
I ran some tests locally and figured out that for the python 3.4 build there are two possible solutions that don't result in a gazillion of slow pip installs:
1. Pin numpy version: `NUMPY_VERSION=1.11`
2. Do not require a specific numpy version: `NUMPY_VERSION=''`

Both solutions result in numpy 1.11.3 being installed.
Which solution do you think is better?

**Comment by zemanj** [comment]
I investigated the issue further:

* Python 3.4 builds fail since the "stable" numpy version in `ci-helpers` have been bumped from 1.14 to 1.15 on September 5. This is because before trying to install other dependencies, `ci-helpers` starts by setting up the test env with
```bash
conda create -n test python=$PYTHON_VERSION numpy=$NUMPY_VERSION
```
and that fails because in the conda-forge channel, there is no python 3.4 build for numpy 1.15.
* Before the stable version bump, the python 3.4 builds only passed because of the pip fallback. So the numerous pip installs we're currently seeing when pinning `NUMPY_VERSION=1.14` are in fact nothing new, they just went unnoticed. See, e.g., [this build](https://travis-ci.org/MDAnalysis/mdanalysis/jobs/417655534) from August 24.
* The biggest problem with pip installs is that they're especially slow for scipy, and the latest python 3.4 build of scipy is version 0.18.1 with numpy 1.11.3. The same is true for biopython.

**Conclusion:**
Having a conda-forge numpy 1.15 build for python 3.4 will *only* help *in conjunction* with builds of scipy and biopython (and maybe further dependencies) for python 3.4 and numpy 1.15.

Until that happens (I don't think it ever will), we can either use solution 1 above (safe) or solution 2 (unsafe, but will automatically raise the numpy version if suitable builds become available).

**Comment by kain88-de** [comment]
> Having a conda-forge numpy 1.15 build for python 3.4 will only help in conjunction with builds of scipy and biopython (and maybe further dependencies) for python 3.4 and numpy 1.15.

Only a numpy build for python 3.4 should be required in theory. The best-practice for packages is to build against numpy 1.8 which is ABI compatible with any later version. MDAnalysis, for example, builds it's conda-forge packages against numpy 1.10 and can be run in combination with any later version.
This can work because we *only* pin numpy and conda is free to choose corresponding versions for our other dependencies.

The simple solution is to drop 3.4 though.  Keep in mind our original argument for python3.4 was that this is the current python version in LTS distributions like CentOS, OpenSuse and Ubuntu LTS (not sure for 2018.4!).  These operating systems are common on HPC clusters.

**Comment by zemanj** [comment]
Dropping testing for Python 3.4 essentially implies dropping support for Python 3.4 entirely, since the functionality can no longer be guaranteed.
> The simple solution is to drop 3.4 though. Keep in mind our original argument for python3.4 was that this is the current python version in LTS distributions like CentOS, OpenSuse and Ubuntu LTS (not sure for 2018.4!). These operating systems are common on HPC clusters.

So dropping Python 3.4 testing is not really an option.

**Comment by kain88-de** [comment]
> So dropping Python 3.4 testing is not really an option.

It is an option. Conda offers a good way to get newer python version on a cluster. I just wanted to mention why we set the oldest python 3 version to 3.4 when we added python 3 support last year. I would also be OK to say we only support the two most recent versions. We only have limited developer time anyway.

**Comment by orbeckst** [comment]
In general I agree with dropping support when we don't have the time to support it. In this case, it seems an easy fix, though, that does not really cost us much, now that @zemanj figured out what is happening.

I think a user with Python 3.4 would do what is equivalent to ``NUMPY_VERSION=''``: just let conda figure it out. I think that's the sensible thing to do. Or what are the downsides?

For downstream packagers it would probably also be easier if we are not too restrictive about Python versions.


(ci-helpers does not document that they pin numpy to a specific version by default. If not, this should be mentioned because of exactly the problem we are having here, which breaks the assumption that you can rely on conda to figure out the dependencies. I understand the rationale of working with a specific version of numpy for a scientific package, though. In any case, I'm happy to open an issue once we know what we're doing.)

**Comment by zemanj** [comment]
I think the only downside is that, at the moment, we won't have python 3.4 tests with the latest numpy version. Maybe still better than nothing, though. Shall I revive PR #2067 or open a new PR?

**Comment by kain88-de** [comment]
We do set the standard numpy version in tests to be *stable* meaning the last stable release. 

https://github.com/MDAnalysis/mdanalysis/blob/b5ba583f8834b384a596691f36a787a3aeeef9f0/.travis.yml#L35

I set this to ensure we **always** test against the latest numpy release and make sure everything works. This works independent of what conda thinks and gives us a *better* reproducible test environment (at least I thought so). Our other tests run against develop and 1.10 our minimum required version.  This combination should ensure we are good from numpy 1.10 to the latest and greatest.

**Comment by zemanj** [comment]
@kain88-de That's exactly what I intended to say with my last comment, thanks for clarifying. If we set `NUMPY_VERSION=''` in the 3.4 tests, this overwrites the global setting, and thus, we lose tests with the latest stable numpy release for python 3.4.

**Comment by tylerjereddy** [comment]
We've removed Python 3.4 from the NumPy CI testing matrices; I think from the upcoming 1.16 there's no "official" 3.4 support, though we're not trying to break compatibility right out of the gate.

**Comment by orbeckst** [comment]
@MDAnalysis/coredevs the motion to drop 3.4 support is on the table (see details in #2066). 

Summary
- latest numpy does not officially test 3.4 anymore
- our tests for 3.4 cannot use the latest numpy because the package dependencies do not resolve properly (we are currently pinning it to an earlier version but that's a band-aid that's also not working properly, details in PR #2067); more importantly, the testing philosophy has been to always test against *latest stable numpy* and the *our oldest support numpy* (with an older Python version). We cannot follow through with this philosophy anymore for 3.4
- developer time is precious, so just supporting Python 2.7 (until 2020) and last two Python releases (3.5, 3.6, ... well 3.7 lingers with PR #1963 ) would be ok

Opinions, simple 👍 or 👎 ?

**Comment by jbarnoud** [comment]
We are supporting older versions of numpy that still support python 3.4. Unless we encounter issues that are specific to python 3.4, I would be in favor of keeping the support and test with an older version of numpy. There is some cost at supporting language features that are absent or different in python 3.4; though it is unlikely that we encounter them as long as we support python 2.7.

I am in favor of dropping support for python 3.4 when we bump the minimal required version of python to a version that does not support python 3.4, when we drop the support for python 2.7, or for MDAnalysis 1.0.

**Comment by orbeckst** [comment]
We seem to have a majority for dropping 3.4 support.

Could I make a compromise suggestion?
* We officially drop support for 3.4 once numpy 1.16 is officially out and they announce that they drop support for 3.4.
* Until then we test against 3.4, but pin numpy to our *minimal version of numpy* (1.10.4 I think) for the 3.4 tests. (This should allow fast conda-based installs and avoid pip). 
* We add something to the docs that outline what we are supporting and roughly what our policy is with respect to supporting Python 2.7 and latest Python 3.x. (We need this for 1.0 anyway.)

**Comment by richardjgowers** [comment]
Yeah pinning to a 3.4 compatible numpy works too, which is what @jbarnoud is saying, so I'm changing my vote to that I guess

**Comment by orbeckst** [comment]
The discussion in PR #2174 is tending towards dropping 3.4 support – mainly to cut down on developer time for debugging stupid version issues that should be sorted out by the packaging system...

**Comment by richardjgowers** [comment]
fixed in #2174 , py 3.4 dropped
